### PR TITLE
refactor the holepunching code

### PR DESCRIPTION
**Pull request by marten-seemann**
This came out of a discussion with @mxinden and @dennis-tra.

This PR splits the hole punching code into two components:
1. A `Service`, which is run on all hosts, independent of their NAT status. This service accepts DCUtR streams and responds to hole punching requests.
2. A `HolePuncher`, which is only run on hosts behind a NAT. This service initiates DCUtR streams when it detects new relayed connections.

Note: This will be merged for the v0.19.0 release.

### CODE REVIEW

File: p2p/protocol/holepunch/holepuncher.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+			log.Debugw("hole punching failed", "peer", rp, "error", err)
+			hp.tracer.ProtocolError(rp, err)
+			return err
+		}
+		synTime := rtt / 2
+		log.Debugf("peer RTT is %s; starting hole punch in %s", rtt, synTime)
+
+		// wait for sync to reach the other peer and then punch a hole for it in our NAT
+		// by attempting a connect to it.
+		timer := time.NewTimer(synTime)
```
Comment (vyzo) [inline p2p/protocol/holepunch/holepuncher.go:147]: why are we creating a timer for every iteration? This is no better than using time.After.

**Review by vyzo — COMMENTED** [review]
this is a very welcome refactor; some questions about our favourite topic, the timer.

File: p2p/protocol/holepunch/holepuncher.go (hunk 1 above)
Comment (marten-seemann) [reply p2p/protocol/holepunch/holepuncher.go:147]: Because a `time.Timer` can be stopped, whereas a `time.After` cannot.

Comment (vyzo) [reply p2p/protocol/holepunch/holepuncher.go:147]: yes but same inefficiency.
We should move the timer out of the loop, defer stop it, and reset for use.

Comment (marten-seemann) [reply p2p/protocol/holepunch/holepuncher.go:147]: Yeah, I don't care about the inefficiency in this case:
1. If the hole punch succeeds, the code only gets executed once.
2. Even if it doesn't, it's 3 timers on the entire connection
Probably not worth getting into trouble with the Reset API.

Comment (vyzo) [reply p2p/protocol/holepunch/holepuncher.go:147]: ok fair enough, lets avoid the foot bazooka.

**Review by vyzo — APPROVED** [review]

**Pull request merged**